### ISSUE TITLE
Thank you & some questions

### ISSUE THREAD

**Issue by forensicmike** [fpgaminer/bigasp-training#3]
Have to echo what the other person who opened an issue said, this has been extremely inspiring and fun to experiment with. Chaining two custom trained models together is something I had never even conceived of, and the use of ELO rankings to do it is very cool as well. I've had really impressive results with the Classifier model portion of this project, though my test loss still seems to floor at around 0.29 no matter what LR, training steps, dropout, weight decay, or optimizer I use. I also had no idea that embeddings could be used in this way, which makes me think it could have direct applications to other modalities as well (such as rating quality of 2 audios or text)

I'm currently experimenting with doing more intelligent manual compares, where it focuses on pitting "undefeated" embeddings against each other specifically, and am considering extending that even further to dial in on ones where the scores are close to 50% confidence, rather than just constantly adding additional embeddings to the dataset.

A couple of questions for you though:
- With the **Ranker** model, I am struggling to get my test loss below 1.0. if I run a single test through the trained model repeatedly I am getting wildly different results from inference to inference (jumping 8 bins or more). Are its outputs supposed to be non-deterministic? 
- Are the training parameters in the script what you rolled with in the long run? I have the impression this may have been more for testing and your production run had a much larger dataset and more training than is suggested in the code on here.
- Can you please help me understand how to interpret the results of the LR finder? Am I looking for a certain spot on the graph?
- What are your thoughts on increasing model complexity? I have been experimenting with adding additional neurons similar to:
```python
def __init__(self, embedding_size, dropout):
    super(QualityClassifier, self).__init__()

    self.ln = nn.LayerNorm(embedding_size)
    self.seq = nn.Sequential(
        nn.Dropout(dropout),
        nn.Linear(embedding_size * 2, embedding_size * 3),
        nn.GELU(),
        nn.Dropout(dropout),
        nn.Linear(embedding_size * 3, embedding_size * 4),
        nn.GELU(),
        nn.Dropout(dropout),
        nn.Linear(embedding_size * 4, embedding_size * 3),
        nn.GELU(),
        nn.Dropout(dropout),
        nn.Linear(embedding_size * 3, 2)
    )
```
but I honestly have no idea what I'm doing here and am relying a lot on copilot, and am not sure if this would have unintended consequences.

Anyhow, thanks in advance for any time you spend on this, and no worries if it's too much to bother with - please feel free to just close the issue if thats the case. I'm very appreciative regardless!

**Comment by fpgaminer** [comment]
> With the Ranker model, I am struggling to get my test loss below 1.0. if I run a single test through the trained model repeatedly I am getting wildly different results from inference to inference (jumping 8 bins or more). Are its outputs supposed to be non-deterministic?

Are you putting the model into `eval` mode? If not `dropout` will be making it random (and making the results of the test worse).

> Are the training parameters in the script what you rolled with in the long run?

Yes, I think so. And I had about 1500 manual head-to-head ratings in the dataset.  That gets expanded slightly in the notebook using augmentations (blurring and compression).

> Can you please help me understand how to interpret the results of the LR finder? Am I looking for a certain spot on the graph?

LR finder is a technique explained in the fastai course(s); I recommend reviewing those for the best explanation. But roughly the graph should have three sections. A flat initial section where the loss isn't doing much. A decreasing section where the loss is accelerating downwards. A final huge spike exploding the loss upward.  The interesting section is that middle one where the loss is accelerating down.  The "best" LR is where that slope is steepest (decreasing quickest), or somewhere around that point.  The exact LR will depend heavily on the model, dataset, batch size, dropout, etc.  But _generally_ it's between 1e-3 and 1e-4.

> What are your thoughts on increasing model complexity?

Experiments are cheap and fast at this scale.  Try it out and see if it helps your use-case!

**Comment by forensicmike** [comment]
Thank you for your detailed and quick reply. You were right, I wasn't calling `model.eval()` and now that I am, it's giving the same result each time. I will definitely read up on the courses but appreciate your explanation anyhow. I agree that experimenting is fast, the only real limitation is my own patience with seeing results. Lots of trial and error. I'll keep at it though, thanks again!
